lib/middleware/remote_user_auth_middleware.rb

```diff
@@ -0,0 +1,53 @@
+require 'securerandom'
+
+class RemoteUserAuthMiddleware
+
+  DefaultUserHeaderName = "Remote-User"
+  DefaultEmailHeaderName = "Remote-Email"
+  DefaultGroupsHeaderName = "Remote-Groups"
+  # if the header specified by user_header_name is provided in the
+  # request, treat the value of that header as a username, and log in as that
+  # user, creating if necessary.
+  # 
+  # This middleware can also be configured to detect headers containing the
+  # user's email, and comma separated group membership (along with a group name
+  # that signifies the created user should be an admin) during the user
+  # creation phase.
+  #
+  def initialize(app, user_header_name = nil, email_header_name = nil, groups_header_name = nil, admin_group = nil, warden_scope = :user)
+    @app = app
+    @user_header_name = (user_header_name || DefaultUserHeaderName).gsub("-", "_").upcase
+    @email_header_name = (email_header_name || DefaultEmailHeaderName).gsub("-", "_").upcase
+    @groups_header_name = (groups_header_name || DefaultGroupsHeaderName).gsub("-", "_").upcase
+    @admin_group = admin_group
+    @warden_scope = warden_scope
+  end
+
+  def call(env)
+    begin
+      username = env['HTTP_' + @user_header_name]
+      # do nothing if no remote-user is presented
+      if username
+        user = User.find_or_create_by!(username: username) do |u|
+          u.requires_no_invitation_code!
+
+          # needed for devise validations
+          u.password = SecureRandom.urlsafe_base64(24 + rand(8))
+          u.email = env['HTTP_' + @email_header_name] || "#{username}@remoteuser.auth"
+          u.admin = !!@admin_group && env['HTTP_' + @groups_header_name].split(",").include?(@admin_group)
+        end
+
+        warden = env['warden']
+        # perform login unless user is already logged in
+        unless warden.user(@warden_scope) == user
+          warden.set_user(user, { :scope => @warden_scope })
+        end
```

Comment:
I have no clue what the `warden` mechanism does., so this might be totally off.

- `user` can be `nil` if not header is passed
- `warden` can be `nil` if it is not in the environment

That gives four cases, could you please verify?

| user | warden | result |
|-|-|-|
| nil | nil | `warden.user` called on  `nil`. Crash? |
| !nil | nil | `warden.user` called on  `nil`. Crash? |
| nil | !nil|   `warden.user` will be set to `nil` |
|!nil | !nil| `warden.user` will be set to logged in user |

Reply:
warden is the authorization framework used by this application: https://github.com/wardencommunity/warden
It is configured by the application, and can't be nil at this point. If it is nil, this entire middleware is broken, and in my opinion it should fail loudly, producing a clear stack trace in the logs, rather than silently stop working.

Additionally, we find or create the user just above- I don't believe `find_or_create_by!` can return nil (validation failure throws in the bang version), but I can add a guard against nil user.

Reply:
Thanks for explaining!